lib/galaxy/workflow/modules.py

```diff
@@ -652,51 +756,281 @@ def get_all_outputs(self, data_only=False):
+    default_default_value = ''
     parameter_type = default_parameter_type
     optional = default_optional
+    default_value = default_default_value
 
     def get_inputs(self):
-        # TODO: Use an external xml or yaml file to load the parameter definition
-        parameter_type = self.state.inputs.get("parameter_type", self.default_parameter_type)
-        optional = self.state.inputs.get("optional", self.default_optional)
-        input_parameter_type = SelectToolParameter(None, XML(
-            '''
-            <param name="parameter_type" label="Parameter type" type="select">
-                <option value="text">Text</option>
-                <option value="integer">Integer</option>
-                <option value="float">Float</option>
-                <option value="boolean">Boolean (True or False)</option>
-                <option value="color">Color</option>
-            </param>
-            '''))
+        parameter_def = self._parse_state_into_dict()
+        parameter_type = parameter_def["parameter_type"]
+        optional = parameter_def["optional"]
+        select_source = dict(name="parameter_type", label="Parameter type", type="select", value=parameter_type)
+        select_source["options"] = [
+            {"value": "text", "label": "Text"},
+            {"value": "integer", "label": "Integer"},
+            {"value": "float", "label": "Float"},
+            {"value": "boolean", "label": "Boolean (True or False)"},
+            {"value": "color", "label": "Color"},
+        ]
+        input_parameter_type = SelectToolParameter(None, select_source)
+        # encode following loop in description above instead
         for i, option in enumerate(input_parameter_type.static_options):
             option = list(option)
             if option[1] == parameter_type:
                 # item 0 is option description, item 1 is value, item 2 is "selected"
                 option[2] = True
                 input_parameter_type.static_options[i] = tuple(option)
-        return OrderedDict([("parameter_type", input_parameter_type),
-                            ("optional", BooleanToolParameter(None, Element("param", name="optional", label="Optional", type="boolean", value=optional)))])
 
-    def get_runtime_inputs(self, **kwds):
-        parameter_type = self.state.inputs.get("parameter_type", self.default_parameter_type)
-        optional = self.state.inputs.get("optional", self.default_optional)
+        parameter_type_cond = Conditional()
+        parameter_type_cond.name = "parameter_definition"
+        parameter_type_cond.test_param = input_parameter_type
+        cases = []
+
+        for param_type in ["text", "integer", "float"]:
+            if param_type == "text":
+                if parameter_type == "text":
+                    default = parameter_def.get("default") or ""
+                else:
+                    default = ""
+                input_default_value = TextToolParameter(None, XML(
```

Comment:
Can we use the new dict source for this now ?

Reply:
Done locally - will push in a second. Eliminates XML and Element imports 😄.